File: backend/trip-service/internal/model/trip.go
package model

import (
	"bus-booking/trip-service/internal/constants"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Trip struct {
	BaseModel
	RouteID       uuid.UUID            `gorm:"type:uuid;not null" json:"route_id" validate:"required"`
	BusID         uuid.UUID            `gorm:"type:uuid;not null" json:"bus_id" validate:"required"`
	DepartureTime time.Time            `gorm:"type:timestamptz;not null" json:"departure_time" validate:"required"`
	ArrivalTime   time.Time            `gorm:"type:timestamptz;not null" json:"arrival_time" validate:"required"`
	BasePrice     float64              `gorm:"type:decimal(10,2);not null" json:"base_price" validate:"required,min=0"`
	Status        constants.TripStatus `gorm:"type:varchar(50);not null;default:'scheduled'" json:"status" validate:"required"`
	IsActive      bool                 `gorm:"type:boolean;not null;default:true" json:"is_active"`

	Route *Route `gorm:"constraint:OnUpdate:CASCADE" json:"route,omitempty"`
	Bus   *Bus   `gorm:"constraint:OnUpdate:CASCADE" json:"bus,omitempty"`
}

func (Trip) TableName() string {
	return "trips"
}

func (t *Trip) BeforeCreate(tx *gorm.DB) error {
	if t.ID == uuid.Nil {
		t.ID = uuid.New()
	}
	return nil
}

type TripSearchRequest struct {
	// Basic search - now all optional
	Origin      *string `form:"origin" json:"origin,omitempty"`
	Destination *string `form:"destination" json:"destination,omitempty"`

	// Advanced filters - time range instead of date
	DepartureTimeStart *string              `form:"departure_time_start" json:"departure_time_start,omitempty"` // Format: ISO8601 or HH:MM
	DepartureTimeEnd   *string              `form:"departure_time_end" json:"departure_time_end,omitempty"`     // Format: ISO8601 or HH:MM
	ArrivalTimeStart   *string              `form:"arrival_time_start" json:"arrival_time_start,omitempty"`     // Format: ISO8601 or HH:MM
	ArrivalTimeEnd     *string              `form:"arrival_time_end" json:"arrival_time_end,omitempty"`         // Format: ISO8601 or HH:MM
	MinPrice           *float64             `form:"min_price" json:"min_price,omitempty" validate:"omitempty,min=0"`
	MaxPrice           *float64             `form:"max_price" json:"max_price,omitempty" validate:"omitempty,min=0"`
	SeatTypes          []constants.SeatType `form:"seat_types" json:"seat_types,omitempty"`
	Amenities          []constants.Amenity  `form:"amenities" json:"amenities,omitempty"`

	// Status filter (for admin)
	Status *string `form:"status" json:"status,omitempty"`

	// Sorting
	SortBy    string `form:"sort_by" json:"sort_by" validate:"omitempty,oneof=price departure_time duration"`
	SortOrder string `form:"sort_order" json:"sort_order" validate:"omitempty,oneof=asc desc"`

	// Pagination
	Page     int `form:"page,default=1" json:"page" validate:"min=1"`
	PageSize int `form:"page_size,default=20" json:"page_size" validate:"min=1,max=100"`
}

type GetTripByIDRequest struct {
	SeatBookingStatus bool `form:"seat_booking_status" json:"seat_booking_status"`
	PreLoadRoute      bool `form:"preload_route" json:"preload_route"`
	PreLoadRouteStop  bool `form:"preload_route_stop" json:"preload_route_stop"`
	PreloadBus        bool `form:"preload_bus" json:"preload_bus"`
	PreloadSeat       bool `form:"preload_seat" json:"preload_seat"`
}

type TripDetail struct {
	ID             uuid.UUID `json:"id"`
	RouteID        uuid.UUID `json:"route_id"`
	BusID          uuid.UUID `json:"bus_id"`
	DepartureTime  time.Time `json:"departure_time"`
	ArrivalTime    time.Time `json:"arrival_time"`
	BasePrice      float64   `json:"base_price"`
	Status         string    `json:"status"` // Raw string value
	AvailableSeats int       `json:"available_seats"`
	TotalSeats     int       `json:"total_seats"`

	Route      *RouteDetail `json:"route,omitempty"`
	Bus        *BusDetail   `json:"bus,omitempty"`
	PriceTiers []PriceTier  `json:"price_tiers,omitempty"`
}

type RouteDetail struct {
	ID              uuid.UUID `json:"id"`
	Origin          string    `json:"origin"`
	Destination     string    `json:"destination"`
	DistanceKm      int       `json:"distance_km"`
	DurationMinutes int       `json:"duration_minutes"`
}

type BusDetail struct {
	ID         uuid.UUID `json:"id"`
	Model      string    `json:"model"`
	BusType    string    `json:"bus_type"` // Raw string value
	TotalSeats int       `json:"total_seats"`
	Amenities  []string  `json:"amenities"` // Raw string values
}

type PriceTier struct {
	SeatType        string  `json:"seat_type"` // Raw string value
	BasePrice       float64 `json:"base_price"`
	PriceMultiplier float64 `json:"price_multiplier"`
	FinalPrice      float64 `json:"final_price"`
	AvailableCount  int     `json:"available_count"`
}

type ListTripsRequest struct {
	PaginationRequest
	IDStrs []string    `form:"ids[]" json:"ids"`
	IDs    []uuid.UUID `form:"-" json:"-"`
}

type TripResponse struct {
	ID            uuid.UUID      `json:"id"`
	RouteID       uuid.UUID      `json:"route_id"`
	BusID         uuid.UUID      `json:"bus_id"`
	DepartureTime time.Time      `json:"departure_time"`
	ArrivalTime   time.Time      `json:"arrival_time"`
	BasePrice     float64        `json:"base_price"`
	Status        string         `json:"status"` // Raw string value
	IsActive      bool           `json:"is_active"`
	Route         *RouteResponse `json:"route,omitempty"`
	Bus           *BusResponse   `json:"bus,omitempty"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
}

type CreateTripRequest struct {
	RouteID       uuid.UUID `json:"route_id" validate:"required"`
	BusID         uuid.UUID `json:"bus_id" validate:"required"`
	DepartureTime time.Time `json:"departure_time" validate:"required"`
	ArrivalTime   time.Time `json:"arrival_time" validate:"required"`
	BasePrice     float64   `json:"base_price" validate:"required,min=0"`
}

type UpdateTripRequest struct {
	DepartureTime *time.Time            `json:"departure_time,omitempty" validate:"omitempty"`
	ArrivalTime   *time.Time            `json:"arrival_time,omitempty" validate:"omitempty"`
	BasePrice     *float64              `json:"base_price,omitempty" validate:"omitempty,min=0"`
	Status        *constants.TripStatus `json:"status,omitempty" validate:"omitempty,oneof=scheduled in_progress completed cancelled"`
	IsActive      *bool                 `json:"is_active,omitempty"`
}

File: frontend/lib/types/trip.ts
// Trip search request parameters
export interface TripSearchParams {
  // Basic search - all optional now
  origin?: string;
  destination?: string;

  // Time range filters (ISO8601 format)
  departure_time_start?: string;
  departure_time_end?: string;
  arrival_time_start?: string;
  arrival_time_end?: string;

  // Other filters
  passengers?: number; // Client-side only, for seat availability
  seat_type?: "standard" | "vip" | "sleeper";
  price_min?: number;
  price_max?: number;
  departure_time_min?: string; // Deprecated, use departure_time_start
  departure_time_max?: string; // Deprecated, use departure_time_end
  amenities?: string[]; // Filter by bus amenities
  bus_type?: string; // Filter by bus type/model
  operator_id?: string;
  sort_by?: "price" | "departure_time" | "arrival_time";
  sort_order?: "asc" | "desc";

  // Admin filters
  status?: string; // Filter by trip status

  // Pagination
  page?: number;
  page_size?: number;
}

/**
 * Generic type for API enum values with display names
 * Backend returns enums as {value, display_name} objects
 */
export interface DisplayValue<T = string> {
  value: T;
  display_name: string;
}

/**
 * Type alias for string-based constant displays (most common case)
 */
export type ConstantDisplay = DisplayValue<string>;

/**
 * Route information from API response
 */
export interface ApiTripRoute {
  id: string;
  origin: string;
  destination: string;
  distance_km: number;
  duration_minutes: number;
}

/**
 * Bus information from API response
 */
export interface ApiBusInfo {
  id: string;
  model: string;
  plate_number?: string;
  bus_type: string; // Raw string value: "standard" | "limousine" | "sleeper"
  total_seats: number;
  amenities: string[]; // Raw string values: ["wifi", "ac", "toilet", ...]
}

/**
 * Trip item from API search response
 */
export interface ApiTripItem {
  id: string;
  route_id: string;
  bus_id: string;
  departure_time: string; // ISO datetime with timezone
  arrival_time: string; // ISO datetime with timezone
  base_price: number;
  status: string; // Raw string value: "scheduled" | "in_progress" | "completed" | "cancelled"
  is_active: boolean;
  available_seats: number;
  total_seats: number;
  route: ApiTripRoute;
  bus: ApiBusInfo;
}

/**
 * Pagination metadata from API
 */
export interface ApiPaginationMeta {
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

/**
 * Generic paginated response structure from backend
 */
export interface ApiPaginatedResponse<T> {
  data: T[];
  meta: ApiPaginationMeta;
}

/**
 * Complete API search response structure
 */
export interface ApiTripSearchResponse {
  data: ApiTripItem[];
  meta: ApiPaginationMeta;
}

// Trip detail from search response
export interface TripDetail {
  id: string;
  route_id: string;
  bus_id: string;
  departure_time: string; // ISO datetime string
  arrival_time: string; // ISO datetime string
  base_price: number;
  status: string;
  available_seats: number;
  total_seats: number;
  duration: string;
  origin: string;
  destination: string;
  distance_km: number;
  bus_model: string;
  bus_plate_number: string;
  bus_amenities: string[];
  operator_id: string;
  operator_name: string;
}

// Trip search response
export interface TripSearchResponse {
  trips: ApiTripItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// Route types
export interface Route {
  id: string;
  origin: string;
  destination: string;
  distance_km: number;
  estimated_minutes: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  route_stops?: RouteStop[]; // Optional as it might not be returned in list views
}

// Bus Seat type from API
export interface SeatStatus {
  is_booked: boolean;
  is_locked: boolean;
}

export interface BusSeat {
  id: string;
  seat_number: string;
  row: number;
  column: number;
  floor: number;
  seat_type: string; // Raw string value: "standard" | "vip" | "sleeper"
  price_multiplier: number;
  is_available: boolean;
  status?: SeatStatus; // Seat booking status from booking service
}

// Seat layout editor types
export type SeatCellType = "seat" | "empty" | "blocked" | "driver";

export interface SeatLayoutCell {
  id?: string; // Seat ID if it's a seat
  type: SeatCellType;
  seatType?: "standard" | "vip" | "sleeper";
  seatNumber?: string;
  priceMultiplier?: number;
  isAvailable?: boolean;
  row: number;
  column: number;
  floor: number;
}

export interface SeatLayoutFloor {
  floor: number;
  rows: number;
  cols: number;
  cells: SeatLayoutCell[][];
}

export interface SeatLayoutConfig {
  busId: string;
  floors: SeatLayoutFloor[];
}

export interface CreateSeatRequest {
  bus_id: string;
  seat_number: string;
  row: number;
  column: number;
  seat_type: "standard" | "vip" | "sleeper";
  price_multiplier?: number;
  floor: number;
}

export interface BulkCreateSeatsRequest {
  bus_id: string;
  seats: CreateSeatRequest[];
}

// Bus types
export interface Bus {
  id: string;
  plate_number: string;
  model: string;
  seat_capacity: number;
  amenities: ConstantDisplay[]; // Backend returns array of {value, display_name}
  is_active: boolean;
  created_at: string;
  updated_at: string;
  seats?: BusSeat[]; // Optional as it might not be returned in list views
}

// Trip types
export interface Trip {
  id: string;
  route_id: string;
  bus_id: string;
  departure_time: string; // ISO datetime string
  arrival_time: string; // ISO datetime string
  base_price: number;
  status: ConstantDisplay; // Backend returns {value, display_name}
  is_active: boolean;
  created_at: string;
  updated_at: string;
  route?: Route; // Populated when fetching with preload
  bus?: Bus; // Populated when fetching with preload
}

// Seat types
export interface Seat {
  id: string;
  bus_id: string;
  seat_code: string;
  seat_type: "standard" | "vip" | "sleeper";
  is_active: boolean;
}

// Seat detail from trip seats response
export interface SeatDetail {
  id: string;
  seat_code: string;
  seat_type: string;
  is_booked: boolean;
  is_locked: boolean;
  price: number;
}

export interface SeatAvailabilityResponse {
  trip_id: string;
  available_seats: number;
  total_seats: number;
  seats: SeatDetail[];
}

// Route Stop types - matches backend model
export interface RouteStop {
  id: string;
  route_id: string;
  stop_order: number;
  stop_type: ConstantDisplay; // Backend returns {value, display_name} for "pickup" | "dropoff" | "both"
  location: string;
  address?: string;
  latitude?: number;
  longitude?: number;
  offset_minutes: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateRouteStopRequest {
  stop_order: number;
  stop_type: string; // "pickup" | "dropoff" | "both"
  location: string;
  address?: string;
  latitude?: number;
  longitude?: number;
  offset_minutes: number;
}

export interface UpdateRouteStopRequest {
  stop_order?: number;
  stop_type?: string;
  location?: string;
  address?: string;
  latitude?: number;
  longitude?: number;
  offset_minutes?: number;
  is_active?: boolean;
}
